administrator/components/com_jedchecker/libraries/rules/language.php

```diff
@@ -0,0 +1,177 @@
+
+
+// Include the rule base class
+require_once JPATH_COMPONENT_ADMINISTRATOR . '/models/rule.php';
+
+
+/**
+ * class JedcheckerRulesLanguage
+ *
+ * This class validates language ini file
+ *
+ * @since  2.3
+ */
+class JedcheckerRulesLanguage extends JEDcheckerRule
+{
+	/**
+	 * The formal ID of this rule. For example: SE1.
+	 *
+	 * @var    string
+	 */
+	protected $id = 'LANG';
+
+	/**
+	 * The title or caption of this rule.
+	 *
+	 * @var    string
+	 */
+	protected $title = 'COM_JEDCHECKER_LANG';
+
+	/**
+	 * The description of this rule.
+	 *
+	 * @var    string
+	 */
+	protected $description = 'COM_JEDCHECKER_LANG_DESC';
+
+	/**
+	 * Initiates the search and check
+	 *
+	 * @return    void
+	 */
+	public function check()
+	{
+		// Find all INI files of the extension (in the format tag.extension.ini or tag.extension.sys.ini)
+		$files = JFolder::files($this->basedir, '^[a-z]{2,3}-[A-Z]{2}\.\w+(?:\.sys)?\.ini$', true, true);
+
+		// Iterate through all the ini files
+		foreach ($files as $file)
+		{
+			// Try to validate the file
+			$this->find($file);
+		}
+	}
+
+	/**
+	 * Reads and validates an ini file
+	 *
+	 * @param   string  $file  - The path to the file
+	 *
+	 * @return bool True on success, otherwise False.
+	 */
+	protected function find($file)
+	{
+		$lines = file($file);
+		foreach ($lines as $lineno => $line)
+		{
+			$line = trim($line);
+
+			// Check for BOM sequence
+			if ($lineno === 0 && strncmp($line, "\xEF\xBB\xBF", 3) === 0)
+			{
+				$this->report->addError($file, JText::_('COM_JEDCHECKER_LANG_BOM_FOUND'), 1);
+				// Romeve BOM for further checks
+				$line = substr($line, 3);
+			}
+
+			// Skip empty lines, comments, and section names
+			if ($line === '' || $line[0] === ';' || $line[0] === '[')
+			{
+				continue;
+			}
+
+			// Report incorrect comment character
+			if ($line[0] === '#')
+			{
+				$this->report->addError($file, JText::_('COM_JEDCHECKER_LANG_INCORRECT_COMMENT') .
+					'<br>' . htmlspecialchars($line), $lineno);
+				continue;
+			}
+
+			// Check for "=" character in the line
+			if (strpos($line, '=') === false)
+			{
+				$this->report->addError($file, JText::_('COM_JEDCHECKER_LANG_WRONG_LINE') .
+					'<br>' . htmlspecialchars($line), $lineno);
+				continue;
+			}
+
+			// Extract key and value
+			list ($key, $value) = explode('=', $line, 2);
+
+			$key = rtrim($key);
+
+			// Check for empty key
+			if ($key === '')
+			{
+				$this->report->addError($file, JText::_('COM_JEDCHECKER_LANG_KEY_EMPTY') .
+					'<br>' . htmlspecialchars($line), $lineno);
+				continue;
+			}
+
+			// Check for spaces in the key name
+			if (strpos($key, ' ') !== false)
+			{
+				$this->report->addError($file, JText::_('COM_JEDCHECKER_LANG_KEY_WHITESPACE') .
+					'<br>' . htmlspecialchars($line), $lineno);
+				continue;
+			}
+
+			// Check for invalid characters (see https://www.php.net/manual/en/function.parse-ini-file.php)
+			if (strpbrk($key, '{}|&~![()^"') !== false)
```

Comment:
I know of many language strings with JS in it... and yes it's not good practice but it will still work in many cases in Joomla. So I would also move this down to a warning instead of an error.

Reply:
Here we check key only, not a content, and those characters are not allowed by PHP's `parse_ini_file` (see link in the comment above that code line).

Reply:
Oops sorry... missed that. Your right :+1: